crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -216,74 +216,139 @@ impl<'db> InnerIntersectionBuilder<'db> {
+                    *self = Self::new();
+                    return;
+                }
             }
+            if let Some(index) = to_remove {
+                self.positive.remove_index(index);
+            }
+
+            let mut to_remove = None;
+            for (index, existing_negative) in self.negative.iter().enumerate() {
+                // S & ~T = Never    if S <: T
+                if new_positive.is_subtype_of(db, *existing_negative) {
+                    *self = Self::new();
+                    return;
+                }
+                // A & ~B = A    if A and B are disjoint
+                if existing_negative.is_disjoint_from(db, new_positive) {
+                    to_remove = Some(index);
+                }
+            }
+            if let Some(index) = to_remove {
+                self.negative.remove_index(index);
+            }
+
+            self.positive.insert(new_positive);
         }
     }
 
     /// Adds a negative type to this intersection.
-    fn add_negative(&mut self, db: &'db dyn Db, ty: Type<'db>) {
+    fn add_negative(&mut self, db: &'db dyn Db, new_negative: Type<'db>) {
         // TODO `Any`/`Unknown`/`Todo` actually should not self-cancel
-        match ty {
-            Type::Intersection(intersection) => {
-                let pos = intersection.negative(db);
-                let neg = intersection.positive(db);
-                self.positive.extend(pos.difference(&self.negative));
-                self.negative.extend(neg.difference(&self.positive));
-                self.positive.retain(|elem| !neg.contains(elem));
-                self.negative.retain(|elem| !pos.contains(elem));
+        match new_negative {
+            Type::Intersection(inter) => {
+                for pos in inter.positive(db) {
+                    self.add_negative(db, *pos);
+                }
+                for neg in inter.negative(db) {
+                    self.add_positive(db, *neg);
+                }
             }
-            Type::Never => {}
             Type::Unbound => {}
-            _ => {
-                if !self.positive.remove(&ty) {
-                    self.negative.insert(ty);
-                };
+            // ~Literal[True] & bool = Literal[False]
+            Type::BooleanLiteral(bool)
+                if self
+                    .positive
+                    .iter()
+                    .any(|pos| *pos == KnownClass::Bool.to_instance(db)) =>
+            {
+                *self = Self::new();
+                self.positive.insert(Type::BooleanLiteral(!bool));
             }
-        }
-    }
+            _ => {
+                let mut to_remove = None;
+                for (index, existing_negative) in self.negative.iter().enumerate() {
+                    // ~S & ~T = ~T    if S <: T
+                    if existing_negative.is_subtype_of(db, new_negative) {
+                        to_remove = Some(index);
```

Comment:
Same comment as in other `to_remove` cases; we can have two different existing negative types, and add a new negative type that is a supertype of both.

This also should be testable now: we have `~IntLiteral[1] & ~IntLiteral[2]` and we add `~int`

Reply:
Added this as a test.